### DISCUSSION TITLE
Applying constraints

### DISCUSSION THREAD

**Discussion by KhaledNabilSharafeldin** [idaholab/moose#17455] [Q&A General] [answered]
Hello,
I am quite new to MOOSE and I have a lot of questions, but I will start with these here.

- 	Is there a way that I can access and manipulate the stiffness matrix before solving for the unknowns?
- 	How are the constraints are enforced? Do you use penalty method?
- 	Is there an implementation for lagrange multiplier that relates the unknown variables to each other? If not, do you have recommendation where to start implementing it?

What I am trying to do is to apply some constraint relations between the unknown variables (displacement), for instance, if I’m solving for u, then I want u^1=u^2+u^3, where the numbers are node numbers. As you can see, this should decrease the problem DOF, but I am not sure exactly how to do so in MOOSE. Some of these constraints are in relation to the face nodes (up and down) and edges nodes and so on.

Thank you in advance
Khaled

**Comment by roystgnr** [comment, marked as answer]
> How are the constraints are enforced? Do you use penalty method?

The most typical forms of constraints (periodic boundary conditions, hanging nodes on adaptively refined meshes, and Dirichlet boundary conditions) are almost always enforced directly; we have a constraint matrix C that turns a Jacobian K into `C^T*K*C` (and consistently modifies the residual, and may add terms corresponding to heterogeneous offsets for non-zero Dirichlet conditions).

That's not as easy to do with nonlinear constraints, though.  For those we typically use a penalty term or a Lagrange multiplier.  For a penalty method, there might be some classes in include/constraints/ that would be helpful - e.g. although the TiedValueConstraint only does a simple equality constraint (made less simple by the fact that it can enforce equality in contact problems that don't line up precisely node to node), the kernel is straightforward enough and you could try adding your own subclass with a nonlinear kernel.  For a Lagrange multiplier method, perhaps the mortar method code would be the best example?

**Reply by roystgnr** [reply]
> As you can see, this should decrease the problem DOF

Almost forgot - I'm afraid this isn't the case for us.  Not only does a penalty method leave DoF count the same and a Lagrange multiplier add additional DoFs, but even with direct enforcement of constraints we don't currently "condense" the linear system afterwards, we typically either leave constraint equations in the matrix rows corresponding to constrained DoFs, or we leave a 1 on the diagonal to keep those DoFs from causing solvers any difficulty and then we calculate the constrained DoF values directly after the solve is done.

**Comment by KhaledNabilSharafeldin** [comment]
Hello @roystgnr Thank you for your reply! Since the time I have posted this I arrived at these questions:
+ I started using `LinearNodalConstraint` for setting up these constraints using penalty method, but I have quite a few questions.
    + This constraint function inherits from `NodalConstraint` which has a **formulation type** parameter that is either `penalty` or `kinematic`. The `kinematic` one seems to be deleting the original residual for the secondary node, then adding in the new `QpJacobian` provided by the `LinearNodalConstraint`. So, I wanted to verify if this is the case; that it is possible to implement the direct method using this parameter while setting the `_penalty` to 1 and using the `weight` as a sign indicator. The total DoF number does not change since I am adding constraint functions in their place.
    + The way I am setting up these constraints is by passing 2 vectors; the `secondary_nodes` vector has only ONE node ID, while the `primary` has multiple node IDs corresponding to it. Is this the correct way to go through it? I just want to confirm this.
    + For the equation of the constraint itself, do all the variables should be on the RHS of the equation i.e., secondary-primary=0 (`u^1-[ u^2+u^3]=0`), or in the form of secondary=primary (`u^1=u^2+u^3`). This will affect the equation signs and I just would like to confirm it. From what I have seen in the `NodalConstraint` and `LinearNodalConstraint`, there’s no sign switching going on, so I think it should be secondary-primary=0.  
+ Do node IDs numbering start from zero or 1? Because when visualizing the results, the global ID number starts from 1, but it seems like my Dirichlet BC are applied on a node ID 1 off from what I set it to. I am planning on using mesh pointers in the future, but it seems like it is going to take some time to get familiar with it and I have more questions about it, so I am not making it a priority right now. I am using `mesh_renumbering` flag to `false` to avoid any problems for now.

Thank you so much for taking the time to read this! Really appreciate your help!

**Reply by roystgnr** [reply]
I'm afraid I'm going to have to punt on the NodalConstraint questions - I'm not personally familiar with those classes, and looking at https://mooseframework.inl.gov/source/constraints/LinearNodalConstraint.html is a bit dismaying; it looks like it was written by someone so deep in a solid mechanics rabbit hole that they weren't even thinking about the existence of a wider variety of variables.

But for the more libMesh-specific question:

> Do node IDs numbering start from zero or 1?

Yes!  Your node IDs can in general be zero-based, or one-based, or zero-based but with big gaps from adaptive mesh coarsening removing nodes, or sparse from communication-free parallel creation of new nodes ... libMesh supports all of that, which is nice if you want to use such capability but less nice if you wish you could assume nobody else is using that support.  Most commonly we use zero-based numbering, initially contiguous from most mesh generators (or from any mesh generator if you *hadn't* disabled node renumbering, which does a sweep at init/reinit times to eliminate any numbering gaps), but some output formats don't support zero as an ID (or don't support gaps in node numbers, or both) so we have to translate node numbers back and forth, and the node number you see in your debugger might not match the one you see in your viz program.

**Reply by hugary1995** [reply]
Just a quick comment here: the NodalConstraint works for general variables not just displacements in mechanics. I won't comment on whether the original developer is deep in the solid mechanics rabbit hole 😂

**Reply by hugary1995** [reply]
But I agree -- some words in the documentation page might be misleading.

**Reply by KhaledNabilSharafeldin** [reply]
well thank you both @roystgnr and @hugary1995 for the quick responses!
What I am trying to actually do is to apply the **minimal kinematic boundary condition**... which seemed like a reasonable thing to do using those classes. I already did this but the "results" seems off, when using the penalty method the solution doesn't really converge, and for the kinematic, it doesn't even solve linearly giving off `DIVERGED_NANORINF` error.
I can see that there is currently ongoing development for this by averaging either the microscopic P or F depending on the applied macroscopic quantities in this #17475 
I am still not sure exactly why the approach mentioned in this question wouldn't work, but I am looking forward for #17475 to be merged.
